Sequence sampler

lvignoli opened the pull request:
Close #342.

A working but unfinished Sequence sampler. The tests fail on purpose: I am still thinking about the best way to test this. Also, not all noise sources have been considered.

A new module called `sampler` exposes the `sampler.sample()` function, which is given a sequence in argument and returns a nested dict of Global and Local samples per basis. The construction of this dict mimics the one in the `simulation` module.

Arguments given to `sampler.sample()` allow for modulation and noise models relevant at the sampling level.

Some comment on the internals:

- `Samples` and `QubitSamples` are two data class to holds and pass the samples around before pasting them to the dict
- A global channel can "decay" to a local one because of random-per-qubit noises or an SLM.
- Global channel, local channel or a global decayed channel are handled a bit differently by `sampler.sample()
- Noises are modeled by functions, see the `NoiseModel` alias ; we expose generators from the `sampler.noises` module to construct the desired noise function

HGSilveri commented:
One first comment from just looking at the file structure: it's important that we keep simulation-specific features contained in the `pulser.simulation` module so that we can later separate it from the rest of the library easily.

lvignoli commented:
The latest commits make the sampler a working feature.

- `QubitSamples` is the only data structure of the module. It holds samples at the ns scale for the amplitude, phase and determinant attached to one qubit. An instance accounts for all the _TimeSlot of one channel or only some of them (or all the _TimeSlot of all channels if even necessary).
- modulation is working
- internals have been refactored and simplified, so that processing of channels is more streamlined regardless their type (global or local).
- there are tests with total coverage

A comment on the design choice: for now I think a function is enough.
It is easy to test (no object to instantiate) and the usage is very simple

```python
sampler.sample(seq)
# just like
json.dumps(obj)
```

Though, I am not satisfied with the way noises are specified. It will be reworked with along with the other features I want to add. So it may evolve towards a more complex implementation.

lvignoli commented:
It should close linked issue #342 and make PR #337 obsolete.

HGSilveri reviewed (commented):
Here is a non-exhaustive review of what you've done so far (I'm still missing most of the tests). 

I would say that overall, there is a tendency to abuse helper functions in a way that makes the code harder to read sometimes. I left a couple comments on that, but I feel that overall the code could use a going over with that in mind.

Other than that, it seems to be in a good shape already!

[Note: Actually, GitHub seems to be crashing so I'll leave these comments for now and follow up with the ones I've lost]

lvignoli commented on pulser/sampler/noises.py:
```diff
@@ -0,0 +1,106 @@
+        noise_amp *= np.exp(-((r / waist_width) ** 2))
+
+        amp = s.amp.copy()
+        amp[np.nonzero(amp)] *= noise_amp
```
Indeed, removed

lvignoli commented on pulser/sampler/noises.py:
```diff
@@ -0,0 +1,106 @@
+
+
+def amplitude(
+    reg: Register, waist_width: float, random: bool = True, seed: int = 0
```
You are right, otherwise it fallbacks to the same random generator with seed 0 if it's not provided, and it's no what we want.

lvignoli commented on pulser/sampler/noises.py:
```diff
@@ -0,0 +1,106 @@
+def compose_local_noises(*functions: NoiseModel) -> NoiseModel:
+    """Helper to compose multiple NoiseModel."""
+    return functools.reduce(
+        lambda f, g: lambda x: f(g(x)), functions, lambda x: x
```
This is a way to write mathematical composition. If `functions = [f, g, h]`, it returns `f∘g∘h`.
The identity `lambda x: x` serves both as an initializer and a default value, so that if `functions` is the empty list or has only one element, it returns the expected result.

lvignoli replied:
I will add details to the docstring.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,298 @@
+        if modulation:
+            s = _modulate(ch, s)
+
+        unmasked_qubits = seq._qids - seq._slm_mask_targets
+        s = [x for x in s if x.qubit in unmasked_qubits]  # SLM
+
+        s = noises.apply(s, ch_noises)
```
Definitely, we should apply the noises before the masking.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,298 @@
+        }
+    else:
+        return {
+            "Global": {
+                basis: new_qty_dict()
+                for basis in ["ground-rydberg", "digital"]
+            },
+            "Local": {
+                basis: new_qdict() for basis in ["ground-rydberg", "digital"]
+            },
```
Indeed, and I found it a bit weird that we could have missing keys so often so I went for a zero arrays.
But in fact the whole purpose of this dictionary output is to be used by the `simulation`, so it's better to stick to the format it expect.

I'll change that.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,298 @@
+) -> dict:
+    """Export the given samples to a nested dictionary."""
+    # Get the duration
+    if not _same_duration(samples):  # pragma: no cover
```
It's defensive coding that may be seen unnecessary.
It felt weird to make it fails on purpose in a test, there are no real way to make this happen by using `sampler.sample()`.

I can remove either the check or write a small test. The test file needs a good refresh in any case.

lvignoli replied:
A test has been added.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,298 @@
+    OTHER = "other"
+
+
+def _key_func(x: _TimeSlot) -> _GroupType:
+    if isinstance(x.type, Pulse) or x.type == "delay":
+        return _GroupType.PULSE_AND_DELAYS
+    else:
+        return _GroupType.OTHER
```
Agreed. It even remove the need for the `_GroupType` enum, since the scope is clearer.

lvignoli commented on pulser/sampler/noises.py:
```diff
@@ -0,0 +1,106 @@
+"""Contains the noise models."""
```
I had understood that it was okay to move to the `sampler` module the sampling-related noises, that is non-stochastic noises related to the atoms-lasers interactions.
I envisioned the noises to be splitted between the module regarding of their types, like so

```
pulser/sampler/noises.py
pulser/simulation/noises.py
```

Sorry if I got it wrong, they may all go to the simulation module as well.

HGSilveri replied:
Yes, I think it's preferable if they are all in the `simulation` module.

lvignoli replied:
Okay, just moved the noises there!

lvignoli commented on pulser/sampler/noises.py:
```diff
@@ -0,0 +1,106 @@
+    return f
+
+
+def doppler(reg: Register, std_dev: float, seed: int = 0) -> NoiseModel:
```
Indeed, `SimConfig` remains the user interface to specify these parameters.

When/if we plug the sampler to the simulation module, `SimConfig` would call these constructor when asked.

HGSilveri replied:
It's a "when" for sure!

lvignoli commented:
@HGSilveri Thanks for your review, the latests commits address your different remarks. You will be pleased to see the noises in the `simulation` module!

I reworked some parts a bit, and dropped a few helper functions that were a bit clumsy by being so short.

I will finalize this with your comments on the tests, especially the ones on the Doppler noise—since it looks to me the doppler noise in the simulation module is a bit "unreal", as Mauro already pointed out last Wednesday.

HGSilveri reviewed (changes requested):
Still a few details to address, but nothing major.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+
+        s = noises.apply(s, ch_noises)
+
+        unmasked_qubits = seq._qids - seq._slm_mask_targets
+        s = [x for x in s if x.qubit in unmasked_qubits]  # SLM
```
I was not fully aware of this, it's fixed.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+        }
+
+    def new_qdict() -> dict:
+        return {qubit: new_qty_dict() for qubit in seq._qids}
```
yes

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+        if addr == "Global":
+            # Take samples on only one qubit and write them
+            a_qubit = next(iter(seq._qids))
+            to_write = [x for x in some_samples if x.qubit == a_qubit]
```
Yes, that's what I do.
The reasoning behind this is that the sampler should be able to provide the info for each qubits independently. 
If the distinction between global and local channels is needed then we handle it at the output time.
It is also easier to handle per qubit noise on global channel by doing so.
 
Besides, this make the `sample()` function very straightforward since there is very little distinction between global and local channels.

For very large register it might becomes a bottle neck, but otherwise I do not think it worths being to clever about this ressource consumption.

HGSilveri replied:
Yes, it's the large registers I'm worried about... It just seems very inefficient, but you may very well be right in assuming it doesn't cause any issues, I just don't know. 

Let's take for example a demanding case with 50 qubits, 3 channels, 10 µs of duration. If I'm not mistaken, this would amount to storing `50 * 3 * 10e3 * 3 = 4.5e6` floats, which amounts to 36 Mb. It seems manageable, so I won't push for cleverness here.

lvignoli replied:
I'll add a comment so that one day we may patch this if we happen to extract for too many qubits.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+    return d
+
+
+def _same_duration(samples: dict[str, list[QubitSamples]]) -> bool:
```
Yes, but now the if branch in `_write_dict` is tested and covered.

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+
+def _sample_slots(N: int, *slots: _TimeSlot) -> list[QubitSamples]:
+    """Gather samples of a list of _TimeSlot in a single Samples instance."""
+    # Same target in one group, guaranteed by the strategy (this seems
+    # weird, it's not enforced by the structure,bad design?)
```
`_sample_slots` expects that `*slots` is a list of `_TimeSlot` that all have the same targets. I hence extract them from only the first element with

```python
qubits = slots[0].targets
```

So I find the overall pattern I have written a bit risky because

1. Nothing prevent to call `_sample_slots()` with slots having different target
2. The correct way to gather `_TimeSlots` is with a `TimeSlotExtractionStrategy` before feeding them to `_sample_slots()`, but this type alias does say nor enforce anything about grouping by same targets. So one may implement a strategy satisfying the prototype, but not respecting the grouping, handing bad samples down the line.

I think it's no big deal for this PR, but it's just a reminder that the design is not satisfactory.

HGSilveri replied:
Ok, I see. Thanks for the clarification!

lvignoli commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,373 @@
+        want = sim.samples
+        got = sample(seq)
+
+        for addr, basis, qty in global_keys:
+            try:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qty], want[addr][basis][qty]
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qty],
+                    np.zeros(len(got[addr][basis][qty])),
+                )
+
+        for addr, basis, qubit, qty in local_keys:
+            try:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qubit][qty],
+                    want[addr][basis][qubit][qty],
+                )
+            except KeyError:
+                np.testing.assert_array_equal(
+                    got[addr][basis][qubit][qty],
+                    np.zeros(len(got[addr][basis][qubit][qty])),
+                )
```
True, both function are testing the same way, so they could enjoy a refactor.

lvignoli commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,373 @@
+    np.testing.assert_array_equal(got, want)
+
+
+def test_amplitude_noise():
```
Sure

lvignoli commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,373 @@
+
+    np.testing.assert_allclose(got, want, atol=1e-2)
+    # Equality at 1e-2 only... Why are they not equal? channel.modulate is
+    # called in both cases.
```
Indeed !

(0.42 is the exact value for the Blackman window)

lvignoli replied:
The test now passes with `np.testing.assert_array_equal`, very nice 👏

HGSilveri replied:
> (0.42 is the exact value for the Blackman window)

You're right, my bad. It just differs because the discrete nature of the samples means that the sum of all the values might not be exactly 0.42.

lvignoli commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,373 @@
+
+    # Exclude the digital basis, since filled with zero vs empty.
+    # it's just a way to check the coherence
+    global_keys = [
+        ("Global", basis, qty)
+        for basis in ["ground-rydberg"]
+        for qty in ["amp", "det", "phase"]
+    ]
+    local_keys = [
+        ("Local", basis, qubit, qty)
+        for basis in ["ground-rydberg"]
+        for qubit in seq._qids
+        for qty in ["amp", "det", "phase"]
+    ]
+
+    for addr, basis, qty in global_keys:
+        np.testing.assert_array_equal(
+            samples[addr][basis][qty], sim.samples[addr][basis][qty]
+        )
+
+    for addr, basis, qubit, qty in local_keys:
+        np.testing.assert_array_equal(
+            samples[addr][basis][qubit][qty],
+            sim.samples[addr][basis][qubit][qty],
+        )
```
With

```python
got = sample(seq)
want = pulser.Simulation(seq).samples
```

you prefer

```python
np.testing.assert_array_equal(got, want)
```

or

```python
np.testing.assert_array_equal(want, got)
```

You wrote the same pattern as I did!

HGSilveri replied:
The point is that, in your pattern, you are predefining the keys in `local_keys` and `global_keys`, while in mine we make sure that every item in `sim.samples` is in `samples`, whatever the keys may be.

lvignoli replied:
Ah okay yeah sure then!

lvignoli commented on pulser/sampler/sampler.py:
```diff
@@ -0,0 +1,279 @@
+    qubits = slots[0].targets
+    amp, det, phase = np.zeros(N), np.zeros(N), np.zeros(N)
+    for s in slots:
+        if type(s.type) is str:  # pragma: no cover
```
`_TimeSlot` of type `retarget` or `delay` are no-ops in this loop: there are no samples to write.

There is nothing to test really, so I skip this branch in the coverage.

lvignoli commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,284 @@
+
+    np.testing.assert_array_equal(got, want)
+
+
```
The reason string says it all.

With the Global to Local decay I have implemented, it's not very easy to recover the same dict structure as the one created in the `simulation` module. Currently there is no way to make a Decayed channel become Global again.

The reason is because of the philosophy behind the internal representation of the samples we currently have. We get the `QubitSamples` per qubit on all channel, and then choose at output time how to represent them, as I explained in another comment.

The addressing of the channel is retained separately. Keeping track of time intervals during which a Global becomes Local only temporarily would need to expand on the way addressing is stored.

lvignoli replied:
A way to resolve this is simply to test against the expected samples rather than the `Simulation.samples` dict, and pass on the fact that all pulses on a global channel after the first one will be written to the local dict rather than the global one.
What do you think?

lvignoli replied:
Sorry the comments are showing at the wrong lines, I refer to `test_SLM()` at line 151.

HGSilveri replied:
I guess you don't need the samples to be exactly the same, as long as they yield equivalent results. So yeah, what you suggested seems reasonable

lvignoli replied:
Ok! I'll make a working test for the SLM by checking against expected samples then, and maybe I'll keep the one against the simulation module to remember the extraction is different—at least as long as the sampler is not used in the simulation module it seems reasonable.

HGSilveri replied:
You keep mentioning "the sampler not being used in the simulation module" - I thought you were doing this sampler to use in the simulation module, is this something you don't want anymore?

lvignoli replied:
Sorry if I sound overly cautious, I definitely see the sampler being used in the simulation module. I simply think it requires a bit more work. I opened issue #349 to track the additional feature the sampler may need to serve this purpose.
In the mean time, I advocate to keep the tests that remind us the differences between the two.

HGSilveri replied:
No need to apologize, I was just confused. I agree, let's do as you say!

HGSilveri commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -0,0 +1,284 @@
+@pytest.mark.xfail(reason=slm_reason)
+def test_SLM():
+    q_dict = {
+        "batman": np.array([-4.0, 0.0]),  # sometimes masked
+        "superman": np.array([4.0, 0.0]),  # always unmasked
```
Hahaha cool 😆

HGSilveri commented on pulser/sampler/sampler.py:
```diff
@@ -1,6 +1,9 @@
 
 It contains many helpers.
+
+NOTE(perf): it not very efficient to hold the same data for all qubit in global
+channels, but i remains manageable for register # with less than 100 qubits.
```
nit: A few typos here, and perhaps this belongs more in a comment than a docstring

HGSilveri commented on pulser/sampler/sampler.py:
```diff
@@ -61,12 +64,10 @@ def sample(
 
         ch_noises = list(common_noises)
 
+        slm_on = len(seq._slm_mask_targets) > 0 and len(seq._slm_mask_time) > 0
```
Empty lists evaluate to `False`, so you can just do:
```suggestion
        slm_on = seq._slm_mask_targets and seq._slm_mask_time
```

lvignoli replied:
Thanks for this nice tip

HGSilveri commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -60,79 +60,30 @@ def test_one_pulse_sampling():
 
 
-def test_sequence_sampler(seq):
+def check_same_samples_as_sim(seq: pulser.Sequence):
```
Much cleaner!

HGSilveri commented on pulser/tests/test_sequence_sampler.py:
```diff
@@ -174,101 +125,59 @@ def test_inXY() -> None:
-        a *= amplitude
-        a *= np.exp(-((r / waist_width) ** 2))
-        return a
+    print(sample(seq))
+    print(pulser.Simulation(seq).samples)
```
I guess these were forgotten here?

lvignoli replied:
Indeed

lvignoli commented:
Apart from the doppler noise effect discussion—which fails on purpose now—it looks nice to me.

lvignoli commented:
🎉
